### ISSUE TITLE
[BUG] ComplementSelection is unimplemented in nightly

### ISSUE THREAD

**Issue by littleredcomputer** [genjax-community/genjax#851]
**Describe the bug**
`selection.complement()` produces a ComplementSelection object, but there are no overrides of `.filter` which accept this type

**To Reproduce**
# copied from a doc snippet
```python
import jax
             import genjax
             from genjax import bernoulli
             console = genjax.console()

             @genjax.Static
             def model():
                 x = bernoulli(0.3) @ "x"
                 y = bernoulli(0.3) @ "y"
                 return x

             key = jax.random.PRNGKey(314159)
             tr = model.simulate(key, ())
             chm = tr.strip()
             selection = genjax.select("x")
             complement = selection.complement()
             filtered = chm.filter(complement)
             print(console.render(filtered))
```
**Expected behavior**
a trace containing `y` but not `x`

**Comment by femtomc** [comment]
@littleredcomputer The problem here is a design one, which we should discuss.

Against the "selection interface", the notion of the complement has automatically defined methods given a base selection (e.g. for `has_addr` -- it's just the negation of calling `has_addr` on the base complement). Similarly for getting choices in the implementation of `filter`.

It seems like there should be some way of defining the `ComplementSelection` type -- such that the overloads work correctly.

Part of this is looking at the current implementations of `filter` and `project` in `Trace` and `Choice` types and making sure that their implementations only use the interfaces that `Selection` exposes`. Then, trivially, if we define `ComplementSelection(Selection)`, it should work.

**Comment by littleredcomputer** [comment]
How about this: we focus on Choices, since Traces can be mapped to Choices via `Trace.get_choices()`.

We say any Choice can supply an iterator over its items, which are (key, value) tuples:
```
EmptyChoice -> []
ChoiceValue -> [((), v)]
IndexedChoiceMap -> [(i1, v1), (i2, v2), ...]
HierarchicalChoiceMap -> [((k11, k12, ...), v1), ((k21, k22, ...), v2), ...]
VectorChoiceMap -> I'm not sure
```
And then say: any selection has a function which maps keys to boolean, regardless of type. Complement(Selection) reverses that choice by wrapping the judgment function. To implement filter, iterate over the items, keeping the ones that pass the test, and create a new resultant value over the pairs that remain.

**Comment by sritchie** [comment]
Fixed as of 0.4.0, release coming shortly.
